### PR TITLE
Add Set terminate Option for user

### PR DESCRIPTION
**Pull request by ajindal1**
Add an option to Terminate the current session and the session remains in the terminated state until the user calls the unset terminate option.

### CODE REVIEW

File: src/models/model.cpp [hunk 1]
```diff
@@ -37,9 +37,9 @@ static std::string CurrentModulePath() {
 
 namespace Generators {
 
-State::State(const GeneratorParams& params, const Model& model)
+State::State(const GeneratorParams& params_, const Model& model)
```
Comment (RyanUnderhill) [inline src/models/model.cpp:40]: Was this rename accidental?

File: src/generators.h [hunk 2]
```diff
@@ -66,6 +66,7 @@ struct GeneratorParams : std::enable_shared_from_this<GeneratorParams>, LeakChec
   int batch_size{1};
   int max_batch_size{0};
   bool use_cuda_graph{};
+  mutable bool session_terminated{false};
```
Comment (RyanUnderhill) [inline src/generators.h:69]:
```suggestion
  mutable bool session_terminated{};
```

File: src/generators.cpp [hunk 3]
```diff
@@ -184,6 +190,9 @@ bool Generator::IsDone() const {
 }
 
 void Generator::GenerateNextToken() {
+  if (state_->params_->session_terminated) {
+    throw std::runtime_error("Session in Terminated state, exiting!");
+  }
```
Comment (RyanUnderhill) [inline src/generators.cpp:195]:
```suggestion
  if (state_->params_->session_terminated)
    throw std::runtime_error("Session in Terminated state, exiting!");
```

File: src/generators.h (hunk 2 above)
Comment (ajindal1) [reply src/generators.h:69]: I like the initialization to be explicit, any harm in adding the false flag?

File: src/models/model.cpp (hunk 1 above)
Comment (ajindal1) [reply src/models/model.cpp:40]: I had some compilation issues with the original code, let me retest it, thanks.

File: src/generators.cpp [hunk 4]
```diff
@@ -154,6 +154,9 @@ Generator::Generator(const Model& model, const GeneratorParams& params) : model_
 }
 
 void Generator::ComputeLogits() {
+  if (state_->params_->session_terminated) {
```
[the remark this answers is not in the document]
Comment (ajindal1) [reply src/generators.cpp]: That's a good idea. Although I couldn't find any common functions file so I implemented this in generators.cpp file or should I implement this somewhere else?

**Review by baijumeswani — COMMENTED** [review]
Is it possible to add a unit test for this feature?

**Comment by baijumeswani** [comment]
Do we intend to expose this in other language bindings?

File: src/generators.h (hunk 2 above)
Comment (RyanUnderhill) [reply src/generators.h:71]: Empty braces are zero/false initialization. We use this everywhere, even in the above code, max_batch_size{0} should also use it. Less to read.

File: src/generators.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (RyanUnderhill) [reply src/generators.cpp]: You can just put it in the generator class as a private method.

File: src/generators.cpp [hunk 5]
```diff
@@ -30,6 +30,11 @@ std::string CurrentModulePath() {
 }
 #endif
 
+void ThrowErrorIfSessionTerminated(bool is_session_terminated) {
+  if (is_session_terminated)
+    throw std::runtime_error("Session in Terminated state, exiting!");
```
[the remark this answers is not in the document]
Comment (ajindal1) [reply src/generators.cpp:35]: Good question, one option I could think of using std::bad_function_call or any different type of exception. Is there anything else in your mind?

File: src/generators.h [hunk 6]
```diff
@@ -66,6 +68,7 @@ struct GeneratorParams : std::enable_shared_from_this<GeneratorParams>, LeakChec
...
```
[the remark this answers is not in the document]
Comment (ajindal1) [reply src/generators.h]: Moved, thanks.

File: src/generators.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (ajindal1) [reply src/generators.cpp:35]: I have added another function (IsSessionTerminated) for the user to check if the session is in terminated state, they can use this to verify. Does that sound good to you?

[the remark this answers is not in the document]
Comment (baijumeswani) [reply src/generators.cpp:35]: I think this works for now. @RyanUnderhill any other ideas?
One other idea is that we could use error codes in our C API that let the user know the reason for the returned error.
But I think having a function to check if the session is terminated is also ok.

[the remark this answers is not in the document]
Comment (RyanUnderhill) [reply src/generators.cpp:35]: Yeah, error codes would be a large change and not gain us much. We can rethink it in the future but I don't think they're worth the complexity cost. The IsTerminated() lets users know.

File: src/models/model.cpp [hunk 7]
```diff
@@ -65,7 +65,18 @@ void State::Run(OrtSession& session, int new_batch_size) {
   }
```
Comment (RyanUnderhill) [inline src/models/model.cpp:65]: In session.Run(...) above on line 58, we don't do anything on termination?

The issue is that there's a potential error race. If the user calls 'SetTerminate()' and then a non terminate error was thrown from session.Run() the user will not know it was a non termination related error (unless they look at the string).

Can we detect termination from session.Run? If so, we should rename session_terminated to session_terminate_set and have a second variable for "session_terminated" for if we actually hit the termination case (or if we call ThrowErrorIfSessionTerminated, as that will catch the non session.Run terminate cases).

@baijumeswani can review my thinking also.

File: test/c_api_tests.cpp [hunk 8]
```diff
@@ -281,6 +283,54 @@ TEST(CAPITests, GetOutputCAPI) {
...
+  try{
+    while (!generator->IsDone()) {
+      generator->ComputeLogits();
+      generator->GenerateNextToken();
+    }
+  }
+  catch (const std::exception& e) {
+    std::cout << "Session Terminated: " << e.what() << std::endl;
+  }
+}
```
[the remark this answers is not in the document]
Comment (ajindal1) [reply test/c_api_tests.cpp]: Done, thanks.

**Comment by ajindal1** [comment]
Since we wanted to add runoptions instead of individual APIs, created a new PR for this work https://github.com/microsoft/onnxruntime-genai/pull/1011. Will close this once the PR is merged.

**Pull request closed without merging**